bastionzero/zli
# MFA Grace Period Integration and UX Improvements

## Description of the change

This PR integrates the MFA grace period feature into the ZLI. The goals were to convey that there is a grace period, to allow the user to defer setting up MFA during their grace period, and to provide a UX that is similar to what was already done in the web app.

Other UX changes:
- I changed/added text in various places to improve the UX or to match what is in the web app
- I changed the browser behavior for MFA scenarios, so that the browser doesn't hang and wait for the ZLI MFA interaction to complete. It seemed potentially confusing to the user that the browser window was blank. It didn't seem intuitive. Telling the user the SSO login was successful and directing them to the ZLI seems like a better UX.
- I changed the QR code to be small. But this only looks good on Windows and Linux, so the small QR code is only shown on those platforms, and the large one is used for OS X.

## Testing

1. Enable MFA for a user and log in using the ZLI. When asked about setting up MFA, type 'n'. After saying no, verify that authentication completed successfully by running another command. Log out and log back in with that user to verify that the same MFA setup question is asked.
2. Enable MFA for the same user and log in using the ZLI. When asked about setting up MFA, type 'y'. After saying yes, verify that authentication completed successfully by running another command. Log out and log back in with that user to verify that the MFA setup did indeed complete by seeing the MFA code prompt instead of the MFA setup flow.
3. Disable MFA for the same user and make sure no MFA actions are presented. 

**backend branch:** 
**bzero branch:** 
**charts branch:** 

#### Ready to run system tests?


## Relevant release note information

Release Notes: Allow for MFA grace period in ZLI

## Related JIRA tickets

Relates to JIRA: CWC-2521

## Relevant HotFix if applicable

Relevant HotFix PR Number:

## Have you considered the security impacts?

Does this PR have any security impact?


If yes, please explain:

Does this PR add a new/modifies the name/arguments of a `zli` command?


If yes, please provide the docs/zli-reference-manual branch reflecting the changes:
